src/lib/device/device_manager.cpp

```diff
@@ -462,348 +48,349 @@ int DeviceManager::get_open_flags(const io_flag oflags) {
+    } else {
+        pdevs.erase(pdevs.begin() + 1, pdevs.end()); // TODO: Pick random one
     }
 
-    auto chunk = std::make_unique< PhysicalDevChunk >(pdev, slot, start_offset, size, c);
-    PhysicalDevChunk* craw = chunk.get();
-    pdev->attach_chunk(craw, prev_chunk);
+    auto input_vdev_size = vparam.vdev_size;
+    vparam.vdev_size = sisl::round_up(vparam.vdev_size, vparam.num_chunks * vparam.blk_size);
+    if (input_vdev_size != vparam.vdev_size) {
+        LOGINFO("{} Virtual device is attempted to be created with size={}, it needs to be rounded to new_size={}",
+                vparam.vdev_name, in_bytes(input_vdev_size), in_bytes(vparam.vdev_size));
+    }
+    uint32_t chunk_size = vparam.vdev_size / vparam.num_chunks;
+
+    LOGINFO(
+        "New Virtal Dev={} of size={} with id={} is attempted to be created with multi_pdev_opts={}. The params are "
+        "adjusted as follows: VDev_Size={} Num_pdevs={} Total_chunks_across_all_pdevs={} Each_Chunk_Size={}",
+        vparam.vdev_name, in_bytes(input_vdev_size), vdev_id, vparam.multi_pdev_opts, in_bytes(vparam.vdev_size),
+        pdevs.size(), vparam.num_chunks, in_bytes(chunk_size));
+
+    // Convert the vparameters to the vdev_info
+    auto buf = hs_utils::iobuf_alloc(vdev_info::size, sisl::buftag::superblk, pdevs[0]->align_size());
+    auto vinfo = new (buf) vdev_info();
+    populate_vdev_info(vparam, vdev_id, pdevs, vinfo);
+
+    // Locate and write the vdev info in the super blk area of all pdevs this vdev will be created on
+    for (auto& pdev : pdevs) {
+        uint64_t offset = hs_super_blk::vdev_sb_offset() + (vdev_id * vdev_info::size);
+        pdev->write_super_block(buf, vdev_info::size, offset);
+    }
 
-    HS_LOG(DEBUG, device, "Creating chunk: {}", chunk->to_string());
-    m_data_chunks[chunk->chunk_id()] = std::move(chunk);
-    ++(m_data_dm_derived.chunk_hdr->num_chunks);
+    // Do a callback for the upper layer to create the vdev instance from vdev_info
+    shared< VirtualDev > vdev = m_vdev_create_cb(*vinfo, false /* load_existing */);
+    m_vdevs[vdev_id] = vdev;
 
-    return craw;
-}
+    // Create initial chunk based on current size
+    for (auto& pdev : pdevs) {
+        std::vector< uint32_t > chunk_ids;
 
-void DeviceManager::remove_chunk(const uint32_t chunk_id) {
-    auto& dm_derived = get_dm_derived();
-    HS_DBG_ASSERT_EQ(dm_derived.chunk_info[chunk_id].is_slot_allocated(), true);
-    dm_derived.chunk_info[chunk_id].set_slot_allocated(false); // Free up the slot for future allocations
-    --(dm_derived.chunk_hdr->num_chunks);
-}
+        // Create chunk ids for all chunks in each of these pdevs
+        for (uint32_t c{0}; c < vparam.num_chunks / pdevs.size(); ++c) {
+            auto chunk_id = m_chunk_id_bm.get_next_reset_bit(0u);
+            if (chunk_id == sisl::Bitset::npos) { throw std::out_of_range("System has no room for additional chunks"); }
```

Comment:
Actually yes. This question made me slightly adjust the order. There was a problem in current code where we write vdev info earlier than create chunk. I moved it after create_chunks(), so vdev will persist only after all chunks are created. In case chunk is failed to create or partially created, homestore reboots and no chunk information is gathered - because bitmap is the last one to be written when chunks are created.